internal/core/application/wallet_service.go

```diff
@@ -880,13 +1123,176 @@ func fetchUnspents(
 		}
 	}
 
+	return unspents, nil
+}
+
+func addUnspents(unspentRepo domain.UnspentRepository, unspents []domain.Unspent) error {
 	return unspentRepo.AddUnspents(context.Background(), unspents)
 }
 
-func groupAddressesInfoByScript(info domain.AddressesInfo) map[string]domain.AddressInfo {
-	group := make(map[string]domain.AddressInfo)
-	for _, i := range info {
-		group[i.Script] = i
+func fetchAndAddUnspents(
+	explorerSvc explorer.Service,
+	unspentRepo domain.UnspentRepository,
+	bcListener BlockchainListener,
+	info domain.AddressesInfo,
+) error {
+	unspents, err := fetchUnspents(explorerSvc, info)
+	if err != nil {
+		return err
+	}
+	if unspents == nil {
+		return nil
+	}
+	go startObserveUnconfirmedUnspents(bcListener, unspents)
+	return addUnspents(unspentRepo, unspents)
+}
+
+func spendUnspents(
+	unspentRepo domain.UnspentRepository,
+	unspentKeys []domain.UnspentKey,
+) (int, error) {
+	return unspentRepo.SpendUnspents(context.Background(), unspentKeys)
+}
+
+func addUnspentsAsync(unspentRepo domain.UnspentRepository, unspents []domain.Unspent) {
```

Comment:
why Async?

Reply:
because it handle errors and logs/warns by itself since it's supposed to be called by a go routine

Reply:
i understand your point, but if not called with go routine its actually sync

Reply:
yes I agree with you but I didn't find a better name for these methods.. do you have any suggestion?